dotCMS/src/main/java/com/dotcms/security/secret/ServiceIntegrationAPIImpl.java

```diff
@@ -0,0 +1,165 @@
+        return hostIdentifier + HOST_SECRET_KEY_SEPARATOR + key.toLowerCase();
+    }
+
+    @VisibleForTesting
+    public ServiceIntegrationAPIImpl(final UserAPI userAPI, final LayoutAPI layoutAPI, final SecretsStore secretsRepository) {
+        this.userAPI = userAPI;
+        this.layoutAPI = layoutAPI;
+        this.secretsStore = secretsRepository;
+    }
+
+    public ServiceIntegrationAPIImpl() {
+        this(APILocator.getUserAPI(), APILocator.getLayoutAPI(), SecretsStore.INSTANCE.get());
+    }
+
+    @Override
+    public List<String> listServiceKeys(final Host host) {
+        return secretsStore.listKeys().stream().filter(s -> s.startsWith(host.getIdentifier()))
+               .map(s -> s.replace(host.getIdentifier() + HOST_SECRET_KEY_SEPARATOR, StringPool.BLANK))
+               .collect(Collectors.toList());
+    }
+
+    @Override
+    public Map<String, List<String>> serviceKeysByHost() {
+
+        Map<String, List<String>> m = secretsStore.listKeys().stream()
+                .map(s -> s.split(HOST_SECRET_KEY_SEPARATOR))
+                .collect(Collectors.groupingBy(strings -> strings[0],
+                        Collectors.mapping(strings -> strings[1], Collectors.toList())));
+
+        return m;
+    }
+
+    @Override
+    public Optional<ServiceSecrets> getSecretForService(final String serviceKey,
+            final Host host, final User user) throws DotDataException {
+        if (userAPI.isCMSAdmin(user) || layoutAPI
+                .doesUserHaveAccessToPortlet(INTEGRATIONS_PORTLET_ID, user)) {
+             Optional<char[]> optionalChars = secretsStore
+                    .getValue(internalKey(serviceKey, host));
+            if (optionalChars.isPresent()) {
+                    //We really don't want to cache the json object to protect the secrets.
+                    //Caching happens on the layers below.
+                    return Optional.of(readJson(optionalChars.get()));
+            } else {
+            //fallback
+                optionalChars = secretsStore
+                        .getValue(internalKey(serviceKey, APILocator.systemHost()));
+                if (optionalChars.isPresent()) {
+                    return Optional.of(readJson(optionalChars.get()));
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public void deleteSecret(final String serviceKey, final String propOrSecretName, final Host host, final User user) {
+
+    }
+
+    @Override
+    public void saveSecret(final String serviceKey, final Tuple2<String, Secret> keyAndSecret, final Host host, final User user) {
+
+    }
+
+    @Override
+    public void saveServiceSecrets(final ServiceSecrets bean, final User user)
+            throws DotDataException, DotSecurityException {
+        saveServiceSecrets(bean, APILocator.systemHost(), user);
+    }
+
+    @Override
+    public void saveServiceSecrets(final ServiceSecrets bean, Host host, final User user)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Parameter 'host' is not assigned and could be declared final](https://app.codacy.com/app/dotCMS/core/pullRequest?prid=4623512)